File: chess-game/chess-game.v3/web/client/chess-app/src/app/pages/game/play/play/models/pieces/rook.ts
import { Square } from "src/app/services/models/chess-game";
import { Side } from "../side";
import { Piece } from "./piece";
import { PieceType } from "./piece_type";

export class Rook extends Piece {
    
    override get can_move_over(): boolean {
        return false;
    }

    constructor(piece_id: string, side: Side) {
        super(piece_id, side, PieceType.Rook);
    }

    override validateMove(from: Square, to: Square): boolean {
        const [delta_file, delta_rank] = this.calculateMoveDeltas(from, to);

        return delta_file == 0 || delta_rank == 0;
    }
}

File: backend/chessapp/domain/pieces/rook.py
from ...domain.pieces.piece import Piece
from ...domain.pieces.piece_type import PieceType
from ...domain.value_objects.side import Side

class Rook(Piece):

    def get_points(self):
        return 50

    def __init__(self, side: Side):
        super().__init__(side, PieceType.Rook)
        self._has_moved = False

    def move(self):
        self._has_moved = True

    def has_moved(self):
        return self._has_moved